### PR TITLE
Allow DDP to handle custom dataclass forward outputs

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92334

Differential Revision: [D42554973](https://our.internmc.facebook.com/intern/diff/D42554973)

### CODE REVIEW

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42554973).

**Review by zhaojuanmao — APPROVED** [review]
Wondering whether we can switch to use treemap at some time of point to make the output data type support more general? it is the same for FSDP as well

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -79,6 +80,11 @@ def _find_tensors(obj):
         return itertools.chain(*map(_find_tensors, obj))
     if isinstance(obj, dict):
         return itertools.chain(*map(_find_tensors, obj.values()))
+    if is_dataclass(obj):
+        return itertools.chain(
+            *map(_find_tensors, [getattr(obj, f.name) for f in fields(obj)])
```
Comment (Skylion007) [inline torch/nn/parallel/distributed.py:85]:
```suggestion
            *map(_find_tensors, (getattr(obj, f.name) for f in fields(obj)))
```

**Comment by awgu** [comment]
> Wondering whether we can switch to use treemap at some time of point to make the output data type support more general? it is the same for FSDP as well

`tree_map()` does not support `dataclass` and `PackedSequence` (which FSDP's `_apply_to_tensor()` does). If we get `tree_map()` to cover a superset, then we should be able to migrate.

**Comment by mrshenli** [comment]
Test failure is irrelevant:

```
======================================================================
ERROR [4.113s]: test_compatible_with_named_optimizer (__main__.TestFSDPOptimState)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 539, in wrapper
    self._join_processes(fn)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 765, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 810, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 663, in run_test
    getattr(self, test_name)()
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 541, in wrapper
    fn()
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 172, in wrapper
    return func(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/distributed/fsdp/test_fsdp_optim_state.py", line 1479, in test_compatible_with_named_optimizer
    state_dicts.append(FSDP._optim_state_dict(model, optim))
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 1734, in _optim_state_dict
    return FullyShardedDataParallel.full_optim_state_dict(
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 1291, in full_optim_state_dict
    optim_state_dict=optim.state_dict(),
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/optim/named_optimizer.py", line 145, in state_dict
    return self._post_state_dict({"state": ret_state, "param_groups": ret_groups})
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/optim/named_optimizer.py", line 276, in _post_state_dict
    FSDP._optim_state_dict_post_hook(self.module, self._optimizer, state_dict)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 1694, in _optim_state_dict_post_hook
    return FullyShardedDataParallel._optim_state_dict_impl(
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 1164, in _optim_state_dict_impl
    return _optim_state_dict(
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/fsdp/_optim_utils.py", line 1393, in _optim_state_dict
    fsdp_osd["param_groups"] = _unflatten_param_groups(
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/fsdp/_optim_utils.py", line 1240, in _unflatten_param_groups
    param_group_params = [
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/fsdp/_optim_utils.py", line 1241, in <listcomp>
    param_key_to_param[flat_param_key]
KeyError: '_fsdp_wrapped_module.net1.0.bias'
```

**Pull request closed without merging**